### PR TITLE
Chore/add unit tests

### PR DESCRIPTION
**Pull request by cans**
This pull request introduces the first bits of comprehensive unit testing for this package.

First commit simply adds requirement files on testing and dev tools. It also lays part of the filesystem hierarchy to store tests.

Second commits add tests for the `heroku3.helpers` module. And to that end also fixes issue martyzz1/heroku3.py#52

Third commit adds tests for the `herku3.models.configvars` module. It is a bit more involved than the first two commits as it also includes some refactoring on that module to make it easier to test, render ConfigVars instances less prone to internal state hijacking and some other issues raised in issue martyzz1/heroku3.py#53. As such it requires more scrutiny.

While I think i represent substantial enough changes for justifying a merge in the master branch, I would obviously not advise including this in any release until broader test coverage is available.

### CODE REVIEW

File: heroku3/helpers.py [hunk 1]
```diff
@@ -30,15 +33,16 @@ def is_collection(obj):
...
-    dict_keys=None,
-    **kwargs):
+              in_dict,
+              strs=None,
+              dates=None,
+              ints=None,
+              objects=None,
+              arrays=None,
+              bools=None,
+              dicts=None,
```
Comment (cans) [inline heroku3/helpers.py]: Made the arguments names more similar to those of class attributes used to declare the models attributes. Makes testing a bit easier.

File: heroku3/helpers.py [hunk 2]
```diff
@@ -101,39 +102,3 @@ def to_python(obj,
     # obj.__cache = in_dict
 
     return obj
-
```
Comment (cans) [inline heroku3/helpers.py]: Removed the function below has it is not used anywhere (raises coverage by a little bit).

File: heroku3/models/__init__.py [hunk 3]
```diff
@@ -41,13 +39,26 @@ def __repr__(self):
...
+        # self._pks is (should be, don't see where this is enforeced) a subset
+        # of the set below, hence there is no need to include it.
+        return (list(cls._arrays.keys()) +
+                cls._bools +
+                cls._dates +
+                cls._dicts +
+                cls._ints +
+                list(cls._map.keys()) +
+                cls._strs
+                )
```
Comment (cans) [inline heroku3/models/__init__.py]: Fixes martyzz1/heroku3.py#52

File: heroku3/models/configvars.py [hunk 4]
```diff
@@ -1,82 +1,108 @@
...
+       configuration.update(safe_configuration)
+
+    """
+
+    def __init__(self, data, app, h=None, **kwargs):
+        # assert app is not None
+        # copy: Ensures caller has no way to mutate our internal state outside
+        # our control
+        self.__data = copy.copy(data)
+        self.__app = app
```
Comment (cans) [inline heroku3/models/configvars.py]: Forced object internal state to be *"private"*.

File: heroku3/models/configvars.py [hunk 5]
```diff
@@ -1,82 +1,108 @@
...
-        try:
-            v = self[key]
-        except KeyError:
-            return False
-        else:
-            if v is None:
-                return False
-            else:
-                return True
+        return key in self.__data
```
Comment (cans) [inline heroku3/models/configvars.py]: Much simpler. Assumes local object state is in sync with the remote state of the configuration though.

File: heroku3/models/configvars.py [hunk 6]
```diff
@@ -1,82 +1,108 @@
...
 
-        del self.data[key]
-        return r.ok
+    @property
+    def _resource(self):
+        return 'apps', self.__app.name, 'config-vars'
 
-    def update(self, newconf):
-        payload = self._h._resource_serialize(newconf)
+    def __patch_config(self, config):
```
Comment (cans) [inline heroku3/models/configvars.py]: Factored boilerplate code for resource access that was copied in 2 or 3 methods.

File: heroku3/models/configvars.py [hunk 7]
```diff
@@ -1,82 +1,108 @@
...
+        payload = self._h._resource_serialize(config)
         r = self._h._http_resource(
             method='PATCH',
-            resource=('apps', self.app.name, 'config-vars'),
+            resource=self._resource,
             data=payload
         )
-
         r.raise_for_status()
```
Comment (cans) [inline heroku3/models/configvars.py]: Just saw that `self._h._htts_resource()` actuall calls `r.raise_for_status()`. So that call is just not useful.

**Comment by martyzz1** [comment]
I'll need to go through this in more detail, But words cannot express how grateful I am to you for this... it has been on my todo for an embaressingly long time :)

**Comment by Alir3z4** [comment]
This is great.
Any idea when this would be merged @martyzz1 ?

**Pull request merged**